shared/js/l10n.js

```diff
@@ -1607,20 +1626,32 @@
       }
     }
 
-    if (!this.ctx.availableLocales.length) {
-      // if there was no availableLanguages meta,
-      // register the default locale only
-      this.ctx.registerLocales(this.ctx.defaultLocale);
+    if (navigator.mozApps && navigator.mozApps.getAdditionalLanguages) {
+      navigator.mozApps.getAdditionalLanguages().then(
+        registerLocales.bind(this));
+    } else {
+      registerLocales.call(this);
     }
     return initLocale.call(this);
   }
 
+  function registerLocales(extraLangs) {
```

Comment:
(I take it that this is a reply to my other comment about `ctx.registerLocales`.)

Yes, I agree.  I recall a conversation we had in November about the core library not having any IO methods and instead using IO methods provided by the platform.  I think this would require deeper changes in how we define our IO module right now (e.g. each IO request would need to know the language of the resource).  It also seems that in your current patch you're making changes in the Locale and Context classes to differentiate between the `app` and `extra` sources.  Would you like to move this into the Bindings code?